### PR TITLE
Added the possibility to throttle incoming RPCs

### PR DESCRIPTION
**Pull request by mdorier**
This PR solves a possible race condition in the `DEFINE_MARGO_RPC_HANDLER` macro and adds the possibility to cap the number of in-flight RPCs to a maximum, configured via the "max_pending_rpcs" field in the Margo JSON configuration (default is 0, meaning no max imposed).

## Fix to DEFINE_MARGO_RPC_HANDLER

This macro is currently calling `__margo_internal_finalize_requested` ([here](https://github.com/mochi-hpc/mochi-margo/blob/main/include/margo.h#L1502)) to check if another RPC currently running has requested margo to shut down. It then calls `__margo_internal_incr_pending` ([here](https://github.com/mochi-hpc/mochi-margo/blob/main/include/margo.h#L1509)) to increment the number of running RPCs. This leaves a small window during which an RPC running in another ES could call `margo_finalize`, complete and finalize the margo instance.

This PR solves this by making the `__margo_internal_incr_pending` return 0 if it correctly incremented the number of pending RPCs, and -1 if finalize was requested, which makes calling `__margo_internal_finalize_requested` unnecessary.

## Throttling beyond a maximum number of RPCs

This PR adds a `max_pending_operations` field in the margo instance, and a condition variable in addition to the existing mutex. If an RPC is received and `pending_operations == max_pending_operations`, the condition variable will make `__margo_internal_incr_pending` block. The condition variable is signaled when the number of pending operations is decreased as existing RPCs terminate.

Note that since `__margo_internal_incr_pending` is called in the context of the RPC handler generated by `DEFINE_MARGO_RPC_HANDLER`, this blocking code is executed within the progress loop, so it will also prevent the progress loop from running. Hence this mechanism is the final throttling mechanism to use when all else (e.g. priority pool, custom schedulers) have failed to properly manage the flow of RPCs.

## Potential problem

I thought about this after posting the PR: because this mechanism will block the progress loop, it can create deadlocks if, for example, RPCs that are executing need the progress loop to be actively running to make progress (e.g. if they have a timer, or if they themselves send RPCs or issue bulk operations). So this might not be very clever in the end... Maybe the right thing to do is, instead of blocking, to make the RPC handler fail with `HG_AGAIN` so that the client is invited to retry the operation later?

### CODE REVIEW

File: src/margo-core.c [hunk 1]
```diff
@@ -1585,20 +1586,32 @@ int __margo_internal_finalize_requested(margo_instance_id mid)
...
 
-void __margo_internal_incr_pending(margo_instance_id mid)
+int __margo_internal_incr_pending(margo_instance_id mid)
 {
-    if (!mid) return;
+    if (!mid) return -2;
+    if (mid->finalize_requested) return -1;
     ABT_mutex_lock(mid->pending_operations_mtx);
-    mid->pending_operations += 1;
+    while (mid->max_pending_operations
```
Comment (carns) [inline src/margo-core.c:1594]: I like the fix to the finalize_requested logic, can that be split out as a separate PR?

We can iterate on the question of how to best throttle RPC volume.  I still would like to unify it somehow at a lower level but it's not clear how yet.

**Review by carns — CHANGES_REQUESTED** [review]
I think we should just split out the correction to the finalize race and merge that for now, keep thinking about the throttling problem as a separate activity.

File: src/margo-core.c (hunk 1 above)
Comment (mdorier) [reply src/margo-core.c:1594]: Ok I did another PR for this.

Regarding the problem of throttling, I asked Jerome and apparently if we make the RPC handler return HG_AGAIN it will be propagated to the sender. So we could use this as a mechanism to limit the number of on-going RPCs. That said, this mechanism would treat all the RPCs as equals, while we may want to limit only some RPCs (typically the long-running ones or the ones with large data transfers), or have different limits for different RPCs.

Comment (carns) [reply src/margo-core.c:1594]: maybe there could be a higher level scheduler that hooks into the rpc handlers and the argobots scheduler so that something has full visibility into that stuff and has the ability to influence both scheduling of handlers that have started and also shed load on those that haven't?

**Comment by mdorier** [comment]
I'm going to close this PR without merging since throttling should now be possible with the new serialization mechanism that allows to send a response with an error back to the client without creating a ULT.
